aptos-move/block-executor/src/executor.rs

```diff
@@ -235,14 +236,14 @@ where
             *worker_idx = (*worker_idx + 1) % post_commit_txs.len();
 
             // Committed the last transaction, BlockSTM finishes execution.
-            if txn_idx as usize + 1 == scheduler.num_txns() as usize {
+            if scheduler.txn_index_right_after(txn_idx).is_none() && scheduler.no_more_txns {
```

Comment:
@sitalkedia 
It's not only the `Scheduler`, but also the `TxnLastInputOutput` that needs a change, if given global indices. (and perhaps other blockstm inner state we haven't identified.)

Also existing blockstm is implemented as a blocking function, which i found a bit awkward when adding message handling.
 
Given perf and other concerns, I think it makes sense if we simply don't touch existing blockstm implementation, and have a separate version for prototyping purpose, where we apply any changes we need?

Reply:
>Also existing blockstm is implemented as a blocking function, which i found a bit awkward when adding message handling

Can you elaborate where and why is this a concern? 

>> Given perf and other concerns, I think it makes sense if we simply don't touch existing blockstm implementation, and have a separate version for prototyping purpose, where we apply any changes we need?

Given we intend to check it in main - I am a bit concerned about the code duplication and maintenance overhead this will entail.

Reply:
For `TxnLastInputOutput`, similarly we can have a `TxnLastInputOutputProvider` interface I guess?

Reply:
In fact a more cleaner way to do this might be to have an interface `TxnStateProvider`, which will be passed both to the scheduler and `TxnLastInputOutput`. The `TxnStateProvider` should basically provide following methods

```
trait TxnStateProvide {

fn get_num_txns
fn get_txn_dependency(index)
fn get_txn_status(index)
fn get_last_input(index)
fn get_last_output(index)
}

struct UnshardedTxnState {
    /// Number of txns to execute, immutable.
    num_txns: TxnIndex,

    /// An index i maps to indices of other transactions that depend on transaction i, i.e. they
    /// should be re-executed once transaction i's next incarnation finishes.
    txn_dependency: Vec<CachePadded<Mutex<Vec<TxnIndex>>>>,
    /// An index i maps to the most up-to-date status of transaction i.
    txn_status: Vec<CachePadded<(RwLock<ExecutionStatus>, RwLock<ValidationStatus>)>>,

       inputs: Vec<CachePadded<ArcSwapOption<TxnInput<K>>>>, // txn_idx -> input.

    outputs: Vec<CachePadded<ArcSwapOption<TxnOutput<T, E>>>>, // txn_idx -> output.

}


Impl TxnStateProvider for UnshardedTxnState {
    ...
}
```

Reply:
That's definitely also needed. Will try implement the above ideas and see how it goes